File: internal/app/repo/midtrans/midtrans.go
package midtrans

import (
	"bytes"
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"mahir-trade-be/internal/app/models"
	"net/http"
	"os"
	"time"

	"go.uber.org/dig"
)

type (
	MidtransRepo interface {
		GeneratePaymentLink(ctx context.Context, req models.MidtransGeneratePaymentLinkRequest) (res models.MidtransGeneratePaymentLinkResponse, err error)
		MidtransCheckStatusTransaction(ctx context.Context, orderID string) (res models.MidtransCheckStatusResponse, err error)
	}

	MidtransRepoImpl struct {
		dig.In
	}
)

func NewMidtransRepo(impl MidtransRepoImpl) MidtransRepo {
	return &impl
}

func generateTokenAuthorization() string {
	clientSecret := os.Getenv("MIDTRANS_CLIENT_SECRET") + ":"
	encoded := base64.StdEncoding.EncodeToString([]byte(clientSecret))

	return "Basic " + encoded
}

func (m *MidtransRepoImpl) GeneratePaymentLink(ctx context.Context, req models.MidtransGeneratePaymentLinkRequest) (res models.MidtransGeneratePaymentLinkResponse, err error) {
	generatePaymentLinkURL := os.Getenv("MIDTRANS_BASE_URL") + "/v1/payment-links"

	requestPayload := models.MidtransGeneratePaymentLinkRequest{
		TransactionDetails: req.TransactionDetails,
		UsageLimit:         1,
		Expiry: models.Expiry{
			StartTime: time.Now().Format("2006-01-02 15:04 -0700"),
			Duration:  1,
			Unit:      "days",
		},
		EnabledPayments: []string{
			"credit_card",
			"bca_klikbca",
			"gopay",
			"permata_va",
			"bca_va",
			"bri_va",
			"bni_va",
			"indomaret",
			"shopeepay",
		},
		ItemDetails:     req.ItemDetails,
		CustomerDetails: req.CustomerDetails,
	}

	payload, err := json.Marshal(requestPayload)
	if err != nil {
		slog.ErrorContext(ctx, "[repo][midtrans][Generate PaymentLink] Error marshal request payload: ", err)
		return res, err
	}

	request, err := http.NewRequest("POST", generatePaymentLinkURL, bytes.NewBuffer(payload))
	if err != nil {
		slog.ErrorContext(ctx, "[repo][midtrans][Generate PaymentLink] Error create new request: ", err)
		return res, err
	}

	request.Header.Set("Authorization", generateTokenAuthorization())
	request.Header.Set("Content-Type", "application/json")

	response, err := http.DefaultClient.Do(request)
	if err != nil {
		slog.ErrorContext(ctx, "[repo][midtrans][Generate PaymentLink] Error sending request: ", err)
		return res, err
	}

	defer response.Body.Close()

	body, err := io.ReadAll(response.Body)
	if err != nil {
		slog.ErrorContext(ctx, "[repo][midtrans][Generate PaymentLink] Error reading response body: ", err)
		return res, err
	}

	err = json.Unmarshal(body, &res)
	if err != nil {
		slog.ErrorContext(ctx, "[repo][midtrans][Generate PaymentLink] Error unmarshal response body: ", err)
		return res, err
	}

	return res, nil
}

func (m *MidtransRepoImpl) MidtransCheckStatusTransaction(ctx context.Context, orderID string) (res models.MidtransCheckStatusResponse, err error) {
	checkStatusURL := os.Getenv("MIDTRANS_BASE_URL") + fmt.Sprintf("/v2/%s/status", orderID)

	request, err := http.NewRequest("GET", checkStatusURL, nil)
	if err != nil {
		slog.ErrorContext(ctx, "[repo][midtrans][CheckStatusTransaction] Error create new request: ", err)
		return res, err
	}

	request.Header.Set("Authorization", generateTokenAuthorization())
	request.Header.Set("Content-Type", "application/json")

	response, err := http.DefaultClient.Do(request)
	if err != nil {
		slog.ErrorContext(ctx, "[repo][midtrans][CheckStatusTransaction] Error sending request: ", err)
		return res, err
	}

	defer response.Body.Close()

	body, err := io.ReadAll(response.Body)
	if err != nil {
		slog.ErrorContext(ctx, "[repo][midtrans][CheckStatusTransaction] Error reading response body: ", err)
		return res, err
	}

	err = json.Unmarshal(body, &res)
	if err != nil {
		slog.ErrorContext(ctx, "[repo][midtrans][CheckStatusTransaction] Error unmarshal response body: ", err)
		return res, err
	}

	return res, nil
}

File: src/repo/midtrans/midtrans.ts
// src/repos/MidtransRepo.ts
import axios from "axios";
import * as dotenv from "dotenv";
import {
  Expiry,
  MidtransCheckStatusResponse,
  MidtransGeneratePaymentLinkRequest,
  MidtransGeneratePaymentLinkResponse,
} from "DTO/midtrans.dto";

dotenv.config();

export class MidtransRepository {
  private baseUrl: string;
  private clientSecret: string;

  constructor() {
    this.baseUrl = process.env.MIDTRANS_BASE_URL || "";
    this.clientSecret = process.env.MIDTRANS_CLIENT_SECRET || "";
  }

  private generateTokenAuthorization(): string {
    const encoded = Buffer.from(this.clientSecret + ":").toString("base64");
    return `Basic ${encoded}`;
  }

  // --- Generate Payment Link ---
  async generatePaymentLink(
    req: MidtransGeneratePaymentLinkRequest
  ): Promise<MidtransGeneratePaymentLinkResponse> {
    const url = `${this.baseUrl}/v1/payment-links`;

    const requestPayload: MidtransGeneratePaymentLinkRequest = {
      transaction_details: req.transaction_details,
      usage_limit: 1,
      expiry: {
        start_time: new Date().toISOString().slice(0, 16).replace("T", " "), // format simplified
        duration: 1,
        unit: "days",
      } as Expiry,
      enabled_payments: [
        "credit_card",
        "bca_klikbca",
        "gopay",
        "permata_va",
        "bca_va",
        "bri_va",
        "bni_va",
        "indomaret",
        "shopeepay",
      ],
      item_details: req.item_details,
      customer_details: req.customer_details,
    };

    try {
      const res = await axios.post<MidtransGeneratePaymentLinkResponse>(
        url,
        requestPayload,
        {
          headers: {
            Authorization: this.generateTokenAuthorization(),
            "Content-Type": "application/json",
          },
        }
      );

      if (res.data.error_messages) {
        throw new Error(
          `Failed to generate payment link: ${res.data.error_messages}`
        );
      }

      return res.data;
    } catch (err: any) {
      throw new Error(
        `[Midtrans][generatePaymentLink] error: ${err.message || err}`
      );
    }
  }

  // --- Check Status Transaction ---
  async midtransCheckStatusTransaction(
    orderID: string
  ): Promise<MidtransCheckStatusResponse> {
    const url = `${this.baseUrl}/v2/${orderID}/status`;

    try {
      const res = await axios.get<MidtransCheckStatusResponse>(url, {
        headers: {
          Authorization: this.generateTokenAuthorization(),
          "Content-Type": "application/json",
        },
      });

      return res.data;
    } catch (err: any) {
      throw new Error(
        `[Midtrans][midtransCheckStatusTransaction] error: ${
          err.message || err
        }`
      );
    }
  }
}
